pinot-plugins/pinot-batch-ingestion/pinot-batch-ingestion-spark-3/src/main/java/org/apache/pinot/plugin/ingestion/batch/spark3/SparkSegmentMetadataPushJobRunner.java

```diff
@@ -59,43 +96,75 @@ public void init(SegmentGenerationJobSpec spec) {
   }
 
   @Override
-  public void run() {
-    //init all file systems
-    List<PinotFSSpec> pinotFSSpecs = _spec.getPinotFSSpecs();
-    for (PinotFSSpec pinotFSSpec : pinotFSSpecs) {
-      PinotFSFactory.register(pinotFSSpec.getScheme(), pinotFSSpec.getClassName(), new PinotConfiguration(pinotFSSpec));
-    }
+  public void run()
+      throws Exception {
+    // Initialize all file systems
+    setupFileSystems();
 
-    //Get outputFS for writing output pinot segments
-    URI outputDirURI;
-    try {
-      outputDirURI = new URI(_spec.getOutputDirURI());
-      if (outputDirURI.getScheme() == null) {
-        outputDirURI = new File(_spec.getOutputDirURI()).toURI();
-      }
-    } catch (URISyntaxException e) {
-      throw new RuntimeException("outputDirURI is not valid - '" + _spec.getOutputDirURI() + "'");
-    }
+    // Get outputFS for writing output pinot segments
+    URI outputDirURI = validateOutputDirURI(_spec.getOutputDirURI());
     PinotFS outputDirFS = PinotFSFactory.create(outputDirURI.getScheme());
-    //Get list of files to process
-    String[] files;
-    try {
-      files = outputDirFS.listFiles(outputDirURI, true);
-    } catch (IOException e) {
-      throw new RuntimeException("Unable to list all files under outputDirURI - '" + outputDirURI + "'");
-    }
 
-    List<String> segmentsToPush = new ArrayList<>();
-    for (String file : files) {
-      if (file.endsWith(Constants.TAR_GZ_FILE_EXT)) {
-        segmentsToPush.add(file);
-      }
-    }
+    // Collect files to process
+    List<String> segmentsToPush = getSegmentsToPush(outputDirFS, outputDirURI);
+
+    // Ensure tableConfigURI is set if missing
+    setupTableConfigURI();
+
+    // Retrieve table config and check for consistent push
+    TableConfig tableConfig =
+        SegmentGenerationUtils.getTableConfig(_spec.getTableSpec().getTableConfigURI(), _spec.getAuthToken());
+    boolean consistentPushEnabled = ConsistentDataPushUtils.consistentDataPushEnabled(tableConfig);
 
+    // Determine push parallelism
     int pushParallelism = _spec.getPushJobSpec().getPushParallelism();
     if (pushParallelism < 1) {
       pushParallelism = segmentsToPush.size();
     }
+
+    // Call the appropriate push helper
+    if (consistentPushEnabled) {
+      handleConsistentPush(segmentsToPush, outputDirURI, pushParallelism);
+    } else {
+      handleNonConsistentPush(segmentsToPush, outputDirFS, outputDirURI, pushParallelism);
+    }
+  }
+
+  private void handleConsistentPush(List<String> segmentsToPush, URI outputDirURI, int pushParallelism)
+      throws Exception {
+    Map<String, String> segmentsToPushUriToTarPathMap =
+        SegmentPushUtils.getSegmentUriToTarPathMap(outputDirURI, _spec.getPushJobSpec(),
+            segmentsToPush.toArray(new String[0]));
```

Comment:
Is there a limit on how many segments can be uploaded in a batch? If there are large number of segments to be uploaded metadata push, endReplace might take time/timeout depending on socket timeout.

Reply:
I have not tested the limits of the API; but I reference https://git.corp.stripe.com/stripe-private-oss-forks/pinot/blob/dc992b43/pinot-plugins/pinot-batch-ingestion/pinot-batch-ingestion-standalone/src/main/java/org/apache/pinot/plugin/ingestion/batch/standalone/SegmentMetadataPushJobRunner.java#L36 (the standalone runner) which currently just uploads everything in 1 (my attempt to replicate its' logic into this spark jobs) but we get some level of parallelism here (depending on if you opt for having executors vs drivers only)